### PR TITLE
Adding v1alpha2 CRs, runs controller locally 

### PR DESCRIPTION
**Pull request by jan-est**
Adding v1alpha2 machine and cloud crs.

### CODE REVIEW

File: 03_launch_mgmt_cluster.sh [hunk 1]
```diff
@@ -33,10 +33,21 @@ CABPKPATH="${M3PATH}/cluster-api-bootstrap-provider-kubeadm"
...
+  CAPBMBRANCH="${CAPBMBRANCH:-master}"
+fi
 
 CAPIREPO="${CAPIREPO:-https://github.com/kubernetes-sigs/cluster-api.git}" 
-CAPIBRANCH="${CAPIBRANCH:-master}"
+
+if [ "${V1ALPHA2_SWITCH}" == true ]; then
+  CAPIBRANCH="${CAPIBRANCH:-release-0.2}"
+else 
+  CAPIBRANCH="${CAPIBRANCH:-master}"
```
Comment (maelk) [inline 03_launch_mgmt_cluster.sh]: CAPI is unused if not running v1alpha2

File: 03_launch_mgmt_cluster.sh [hunk 2]
```diff
@@ -136,15 +147,22 @@ function apply_bm_hosts() {
...
+        nohup make run >> capbm.out.log 2>> capbm.err.log &
+      else 
+        touch capbm.out.log
+        touch capbm.err.log
+        make deploy
+        kubectl scale statefulset cluster-api-provider-baremetal-controller-manager -n metal3 --replicas=0
+        nohup make run >> capbm.out.log 2>> capbm.err.log &
+      fi
     else
       make deploy
```
Comment (maelk) [inline 03_launch_mgmt_cluster.sh]: make deploy in v1alpha2 is problematic, please use make install/run for now as we don't have the proper registry and images in place

File: 03_launch_mgmt_cluster.sh [hunk 3]
```diff
@@ -176,7 +194,8 @@ sudo su -l -c 'minikube ssh sudo ip addr add 172.22.0.2/24 dev eth2' "${USER}"
 launch_baremetal_operator
 apply_bm_hosts
 launch_cluster_api_provider_baremetal
-if [ "${V1ALPHA2_SWITCH}" == true ]; then
+
+if [ "${V1ALPHA2_SWITCH}" == true ] && [ "${CAPBM_RUN_LOCAL}" = true ]; then
```
Comment (maelk) [inline 03_launch_mgmt_cluster.sh]: remove this if not using deploy (see other change)

File: create_machine.sh [hunk 4]
```diff
@@ -41,4 +41,4 @@ else
 fi
 ./user_data.sh "${MACHINE_NAME}" "${OS_TYPE}" | kubectl apply -n metal3 -f -
 
-make_machine | kubectl apply -n metal3 -f -
+make_machine | kubectl apply -n metal3 -f -
```
Comment (maelk) [inline create_machine.sh]: You have made a CR file (machine.yaml) for this (please place it in v1alpha1 folder), you can use it in the make_machine function

File: crs/cluster.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+---
```
Comment (maelk) [inline crs/cluster.yaml]: maybe we should create a v1alpha2 folder to contain all the v1alpha2 CRs

File: crs/controlplane.yaml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    name: ${CLUSTER_NAME}-controlplane-0
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+kind: BareMetalMachine
+metadata:
+  name: ${CLUSTER_NAME}-controlplane-0
+spec:
+  image:
+    url: "http://172.22.0.1/images/rhcos-ootpa-latest.qcow2"
+    checksum: "97830b21ed272a3d854615beb54cf004"
```
Comment (maelk) [inline crs/controlplane.yaml]: This needs to be changed to be parameterized

File: create_v1alpha2_machine.sh [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+make_machine() {
+    if [ "${MACHINE_TYPE}" == controlplane ]; then
+        envsubst '${CLUSTER_NAME} ${KUBERNETES_VERSION} ${IMAGE_URL} ${IMAGE_CHECKSUM}' < ${CR_PATH}${CONTROLPLANE_YAML} | cat
+    fi
+    if [ "${MACHINE_TYPE}" == worker ]; then
+        envsubst '${CLUSTER_NAME} ${KUBERNETES_VERSION} ${IMAGE_URL} ${IMAGE_CHECKSUM}' < ${CR_PATH}${WORKER_YAML} | cat
+    fi 
+}
+
+make_machine | kubectl apply -n metal3 -f -
```
Comment (kashifest) [inline create_v1alpha2_machine.sh]: Add a newline

File: crs/cluster.yaml [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    name: ${CLUSTER_NAME}
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+kind: BareMetalCluster
+metadata:
+  name: ${CLUSTER_NAME}
+# spec:
+#   cloudName: ${CLUSTER_NAME}
+#   cloudsSecret:
+#     name: cloud-config
```
Comment (kashifest) [inline crs/cluster.yaml]: Newline

File: crs/controlplane.yaml [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+      name: '{{ ds.meta_data.hostname }}'
+      kubeletExtraArgs:
+        cloud-provider: baremetal
+  clusterConfiguration:
+    apiServer:
+      extraArgs:
+        cloud-provider: baremetal
+    controllerManager:
+      extraArgs:
+        cloud-provider: baremetal
```
Comment (kashifest) [inline crs/controlplane.yaml]: Newline

File: crs/machine.yaml [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  providerSpec:
+    value:
+      apiVersion: "baremetal.cluster.k8s.io/v1alpha1"
+      kind: "BareMetalMachineProviderSpec"
+      image:
+        url: ${IMAGE_URL}
+        checksum: ${IMAGE_CHECKSUM}
+      userData:
+        name: ${MACHINE_NAME}-user-data
+        namespace: metal3
```
Comment (kashifest) [inline crs/machine.yaml]: Newline

File: crs/machinedeployment.yaml [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+metadata:
+  name: ${CLUSTER_NAME}-md-0
+spec:
+  template:
+    spec:
+      joinConfiguration:
+        nodeRegistration:
+          name: '{{ ds.meta_data.hostname }}'
+          kubeletExtraArgs:
+            cloud-provider: baremetal
```
Comment (kashifest) [inline crs/machinedeployment.yaml]: Newline

File: 03_launch_mgmt_cluster.sh [hunk 12]
```diff
@@ -33,10 +33,19 @@ CABPKPATH="${M3PATH}/cluster-api-bootstrap-provider-kubeadm"
...
+else 
+  CAPBMBRANCH="${CAPBMBRANCH:-master}"
+fi
 
 CAPIREPO="${CAPIREPO:-https://github.com/kubernetes-sigs/cluster-api.git}" 
-CAPIBRANCH="${CAPIBRANCH:-master}"
+
+if [ "${V1ALPHA2_SWITCH}" == true ]; then
+  CAPIBRANCH="${CAPIBRANCH:-release-0.2}"
+fi
```
Comment (maelk) [inline 03_launch_mgmt_cluster.sh]: that's ok if line 75 to 98 are conditioned similarly

File: lib/common.sh [hunk 13]
```diff
@@ -72,6 +72,19 @@ export OPENSTACK_CONFIG=$HOME/.config/openstack/clouds.yaml
...
+export CLUSTER_NAME=${CLUSTER_NAME:-"test1"}
+export KUBERNETES_VERSION=${KUBERNETES_VERSION:-"v1.16.0"}
+
+# Image url and checksum
+export IMAGE_NAME=${2:-CentOS-7-x86_64-GenericCloud-1901.qcow2}
+export IMAGE_URL=http://172.22.0.1/images/${IMAGE_NAME}
+export IMAGE_CHECKSUM=http://172.22.0.1/images/${IMAGE_NAME}.md5sum
+
+#Path to CRs
+export V1ALPHA1_CR_PATH=./crs/v1alpha1/
```
Comment (maelk) [inline lib/common.sh]: unused

File: 03_launch_mgmt_cluster.sh [hunk 14]
```diff
@@ -127,7 +139,7 @@ function make_bm_hosts() {
 
 function apply_bm_hosts() {
     list_nodes | make_bm_hosts > bmhosts_crs.yaml
-    kubectl apply -f bmhosts_crs.yaml -n metal3
+    kubectl apply -f bmhosts_crs.yaml -n metal3envsubst '${CLUSTER_NAME} ${KUBERNETES_VERSION}' < ${V1ALPHA2_CR_PATH}${CLUSTER_YAML} | cat
```
Comment (maelk) [inline 03_launch_mgmt_cluster.sh]: there is a problem here

File: 03_launch_mgmt_cluster.sh [hunk 15]
```diff
@@ -176,6 +194,7 @@ sudo su -l -c 'minikube ssh sudo ip addr add 172.22.0.2/24 dev eth2' "${USER}"
 launch_baremetal_operator
 apply_bm_hosts
 launch_cluster_api_provider_baremetal
```
Comment (maelk) [inline 03_launch_mgmt_cluster.sh]: This should be last, the controller does not start properly otherwise, since the capi CRDs do not exist

File: 03_launch_mgmt_cluster.sh [hunk 16]
```diff
@@ -135,7 +147,13 @@ function apply_bm_hosts() {
 #
 function launch_cluster_api_provider_baremetal() {
     pushd "${CAPBMPATH}"
-    if [ "${CAPBM_RUN_LOCAL}" = true ]; then
+    
+    if [ "${V1ALPHA2_SWITCH}" == true ]; then
+      touch capbm.out.log
+      touch capbm.err.log
+      make install 
+      nohup make run >> capbm.out.log 2>> capbm.err.log &
```
Comment (russellb) [inline 03_launch_mgmt_cluster.sh]: This doesn't seem related, at least based on the PR title

File: create_v1alpha2_machine.sh [hunk 17]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (russellb) [inline create_v1alpha2_machine.sh]: Maybe we should have `v1alpha1` and `v1alpha2` directories for these scripts, to keep them more clearly separated?  I know you put `v1alpha2` in the name of another one, but what do you think about directories instead?

Comment (jan-est) [reply create_v1alpha2_machine.sh]: Good idea! I will add new scripts directory and directories for both v1alpha1 and v1alpha2

File: 03_launch_mgmt_cluster.sh [hunk 18]
```diff
@@ -26,18 +26,25 @@ export GOPATH
...
+fi
+
+CAPIREPO="${CAPIREPO:-https://github.com/kubernetes-sigs/cluster-api.git}" 
+
+if [ "${V1ALPHA2_SWITCH}" == true ]; then
+  CAPIBRANCH="${CAPIBRANCH:-release-0.2}"
+fi
+
+CABPKREPO="${CABPKREPO:-https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm.git}" 
 CABPKBRANCH="${CABPKBRANCH:-master}"
```
Comment (maelk) [inline 03_launch_mgmt_cluster.sh]: Is this used if you remove the clone step ?

**Comment by russellb** [comment]
Since this includes a move of the v1alpha1 script, can you prepare a doc patch to the metal3-io.github.io repo that updates the path to the script in the "try-it" doc page?

Otherwise, I'm fine with this going in assuming tests still pass

**Comment by maelk** [comment]
/approve cancel
The v1alpha1 is broken

**Comment by russellb** [comment]
/approve

don't forget to update the script paths in the try-it docs if you haven't already (metal3-io.github.io repo)
